dali/operators/imgcodec/image_decoder.h

```diff
@@ -0,0 +1,785 @@
+
+    st.image_info.region.ndim = roi.use_roi() ? 2 : 0;
+    if (roi.use_roi()) {
+      st.image_info.region.start[0] = roi.begin[0];
+      st.image_info.region.start[1] = roi.begin[1];
+      st.image_info.region.end[0] = roi.end[0];
+      st.image_info.region.end[1] = roi.end[1];
+      decode_out_sh[0] = roi.end[0] - roi.begin[0];
+      decode_out_sh[1] = roi.end[1] - roi.begin[1];
+    }
+
+    // At the moment we are not dealing with floating point outputs in nvimagecodec
+    assert(IsIntegral(st.parsed_sample.orig_dtype));
+
+    int precision = st.image_info.plane_info[0].precision;
+    if (precision == 0)
+      precision = PositiveBits(st.parsed_sample.orig_dtype);
+    bool need_dynamic_range_scaling =
+        NeedDynamicRangeScaling(precision, st.parsed_sample.orig_dtype);
+    st.dyn_range_multiplier = need_dynamic_range_scaling ?
+                                  DynamicRangeMultiplier(precision, st.parsed_sample.orig_dtype) :
+                                  1.0f;
+
+    st.need_processing = st.req_img_type != st.orig_img_type || st.req_layout != st.orig_layout ||
+                         dtype_ != st.parsed_sample.orig_dtype || need_dynamic_range_scaling;
+
+    st.image_info.buffer_kind = std::is_same<MixedBackend, Backend>::value ?
+                                    NVIMGCODEC_IMAGE_BUFFER_KIND_STRIDED_DEVICE :
+                                    NVIMGCODEC_IMAGE_BUFFER_KIND_STRIDED_HOST;
+    int64_t image_buffer_size =
+        volume(decode_out_sh) * TypeTable::GetTypeInfo(st.parsed_sample.orig_dtype).size();
+    st.image_info.buffer_size = image_buffer_size;
+
+    st.decode_out_cpu = {};
+    st.decode_out_gpu = {};
+    if (st.need_processing) {
+      if constexpr (std::is_same<MixedBackend, Backend>::value) {
+        st.device_buf = mm::alloc_raw_async_unique<uint8_t, mm::memory_kind::device>(
+            image_buffer_size, st.image_info.cuda_stream, st.image_info.cuda_stream);
+        st.image_info.buffer = st.device_buf.get();
+        st.decode_out_gpu = {st.image_info.buffer, decode_out_sh, st.parsed_sample.orig_dtype};
+      } else {
+        st.pinned_buf = mm::alloc_raw_unique<uint8_t, mm::memory_kind::pinned>(image_buffer_size);
+        st.image_info.buffer = st.pinned_buf.get();
+        st.decode_out_cpu = {st.image_info.buffer, decode_out_sh, st.parsed_sample.orig_dtype};
+      }
+    } else {
+      st.image_info.buffer = static_cast<uint8_t *>(out.raw_mutable_data());
+    }
+
+    st.image_info.num_planes = 1;
+    st.image_info.plane_info[0].row_stride =
+        decode_out_sh[1] * nchannels * TypeTable::GetTypeInfo(st.parsed_sample.orig_dtype).size();
+    st.image_info.plane_info[0].height = decode_out_sh[0];
+    st.image_info.plane_info[0].width = decode_out_sh[1];
+    st.image_info.plane_info[0].num_channels = nchannels;
+    st.image = NvImageCodecImage::Create(instance_, &st.image_info);
```

Comment:
Just so I understand: we set the image_info that describes how the image will be decoded, right?

Reply:
Yes, image_info is a input/output structure. Some info can be filled during parsing, but the user can also request something differently (e.g. Planar vs. interleaved), and the user needs to provide the buffer there too. So here we are telling nvimagecodec where we want our image to be stored.